pil-analyzer/src/traits_resolver.rs

```diff
@@ -1,131 +1,85 @@
-                    }
-                })
-                .collect::<Vec<_>>()
-        };
+        let mut solved_impls = HashMap::new();
 
-        for (_, (_, def)) in self.definitions.iter() {
-            if let Some(FunctionValueDefinition::Expression(TypedExpression { e: expr, .. })) = def
-            {
-                for (key, value) in resolve_references(expr) {
-                    result
-                        .entry(key.0)
-                        .or_insert_with(HashMap::new)
-                        .insert(key.1, value);
-                }
-            }
-        }
-
-        for identity in self.identities.iter() {
-            for selector in identity
-                .left
-                .selector
-                .iter()
-                .chain(identity.right.selector.iter())
-            {
-                for (key, value) in resolve_references(selector) {
-                    result
-                        .entry(key.0)
-                        .or_insert_with(HashMap::new)
-                        .insert(key.1, value);
-                }
-            }
-            for (key, value) in resolve_references(identity.left.expressions.as_ref()) {
-                result
-                    .entry(key.0)
+        for ref_poly in references {
+            if let Some(solved) = self.resolve_trait_function_reference(&ref_poly, trait_impls) {
+                let (key, type_args) = solved.0;
+                solved_impls
+                    .entry(key)
                     .or_insert_with(HashMap::new)
-                    .insert(key.1, value);
-            }
-            for (key, value) in resolve_references(identity.right.expressions.as_ref()) {
-                result
-                    .entry(key.0)
-                    .or_insert_with(HashMap::new)
-                    .insert(key.1, value);
+                    .insert(type_args, solved.1);
             }
         }
 
-        result
+        solved_impls
     }
-    fn resolve_reference_trait(
+
+    fn resolve_trait_function_reference(
         &self,
         reference: &PolynomialReference,
         trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
     ) -> Option<SolvedImpl> {
-        let name = reference.name.clone();
-        let (trait_decl_name, trait_fn_name) = name.rsplit_once("::")?;
+        let (trait_decl_name, trait_fn_name) = reference.name.rsplit_once("::")?;
         if let Some(impls) = trait_impls.get(trait_decl_name) {
+            let type_args = reference.type_args.as_ref().unwrap().to_vec();
             for impl_ in impls.iter() {
-                let Type::Tuple(TupleType { ref items }) = impl_.type_scheme.ty else {
+                let Type::Tuple(TupleType { items: _ }) = impl_.type_scheme.ty else {
                     unreachable!()
                 };
-                let type_args = reference.type_args.as_ref().unwrap().to_vec();
+
+                let tuple_args = Type::Tuple(TupleType {
```

Comment:
Can you assert that `tuple_args.is_concrete_type()`? At some point we need to implement this for generic types, but we can do it later.